### ISSUE TITLE
Infinite repetition bug still exists

### ISSUE THREAD

**Issue by 99991** [QwenLM/Qwen3-VL#1611]
In some cases, Qwen3-VL and Qwen2.5-VL become stuck in an infinite loop, repeating the same text over and over. Increasing the repetition penalty is not an acceptable solution because it breaks the transcription of naturally repetitive text, for example in tables.

#### Output

    ```markdown
    # NLP Pipeline

    In the following, the NLP pipeline and its components are discussed in depth:

    ```
    Text
      |
      v
    Language identification
      |
      v
    String -> Tokenization -> Sentence splitter -> POS tagger
      |                     |
      |                     v
      |                Sentences
      |                Tokens
      |
      |-----------------|
      | Segmentation    |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |                 |
      |

#### Code

```python
from transformers import Qwen3VLForConditionalGeneration, AutoProcessor
import torch
torch.manual_seed(0)

model = Qwen3VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen3-VL-4B-Instruct",
    dtype="float16",
    device_map="auto"
)

processor = AutoProcessor.from_pretrained("Qwen/Qwen3-VL-4B-Instruct")

# Download image from github and rename to test.png
image_path = "test.png"

messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image",
                "image": image_path,
            },
            {
                "type": "text",
                "text": "Transcribe this image as markdown.",
            },
        ],
    }
]

# Preparation for inference
inputs = processor.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    return_dict=True,
    return_tensors="pt"
)
inputs = inputs.to(model.device)

# Inference: Generation of the output
generated_ids = model.generate(**inputs, max_new_tokens=256)
generated_ids_trimmed = [
    out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
]
output_text = processor.batch_decode(
    generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
)

print(output_text[0])
```

#### test.png

<img width="756" height="567" alt="Image" src="https://github.com/user-attachments/assets/5effe1f1-c1a9-4f06-a58b-a093b6dfa433" />

#### Related

Previous issue: https://github.com/QwenLM/Qwen3-VL/issues/241 (closed without comment)

**Comment by attashe** [comment]
Same issue

**Comment by ucaslcl** [comment]
**Please use “torch_dtype=torch.bfloat16“ to load model, try again~**

<img width="1354" height="1566" alt="Image" src="https://github.com/user-attachments/assets/3f620db6-b4ec-4140-9b86-37798c7becf2" />

**Comment by 99991** [comment]
> Please use “torch_dtype=torch.bfloat16“ to load model, try again~

I added `torch_dtype=torch.bfloat16` to the `Qwen3VLForConditionalGeneration.from_pretrained`, but still get the infinite repeat bug (code below).

I also get the warning `torch_dtype` is deprecated! Use `dtype` instead!`. Are you perhaps using an older PyTorch version or CUDA driver? I am currently using torch 2.9.0 and CUDA 560.35.05.

Also, `torch.bfloat16` is broken on many GPUs (e.g. V100), see this issue: https://github.com/QwenLM/Qwen3-VL/issues/182

I also tried `torch.float32`, but still get infinite repeat bug.

```python
from transformers import Qwen3VLForConditionalGeneration, AutoProcessor
import torch
torch.manual_seed(0)

model = Qwen3VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen3-VL-4B-Instruct",
    torch_dtype=torch.bfloat16,
    device_map="auto"
)

processor = AutoProcessor.from_pretrained("Qwen/Qwen3-VL-4B-Instruct")

messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image",
                "url": "https://raw.githubusercontent.com/99991/testing/refs/heads/master/misc/test.png",
            },
            {
                "type": "text",
                "text": "Transcribe this image as markdown.",
            },
        ],
    }
]

# Preparation for inference
inputs = processor.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    return_dict=True,
    return_tensors="pt"
)
inputs = inputs.to(model.device)

# Inference: Generation of the output
generated_ids = model.generate(**inputs, max_new_tokens=256)
generated_ids_trimmed = [
    out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
]
output_text = processor.batch_decode(
    generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
)

print(output_text[0])
```

**Comment by dwt** [comment]
We see the same issue running those models in Ollama or LM-Studio. I haven't tried increasing the repeat penalty yet, but we do have quite some documents with large repeating sections (with small differences, like tables with almost identical lines but different dates).

Should the repeat penalty be a solution here? Is that an otherwise known issue of the model, or is it likely that the quantizations we use are just too strong?

**Comment by scouzi1966** [comment]
I have the same issue-especially with lower quants of 4 bits. For example "generate a fully functional scientific calculator in html/css/js" will lead to Qwen3V repeating buttons forever. This is on llama.cpp (GGUF) on MacOS but also happens on Apple MLX converted models.

**Comment by ncphu** [comment]
I faced the same issue. Does anyone have final solution?

**Comment by scouzi1966** [comment]
At least we know it's a common model issue and not our code. I will try to mitigate it. 

### **Sill a big thanks to the Qwen team for this awesome model.**

**Comment by 99991** [comment]
I think the problem lies in the way that the Qwen models were trained. When the same token has been generated a few times, the model is stuck in a loop and can not get out. That is true even for the biggest Qwen-models. For example, the model from https://chat.qwenlm.ai/ fails for the following simple image, which has a few repeating zeros, followed by 459. Once it has started spitting out zeros, it just keeps on going forever and ever and never generates the last thee non-zero digits 459.

#### Input image

<img width="738" height="50" alt="Image" src="https://github.com/user-attachments/assets/5f218520-f609-47d2-a5ed-91fadfdd27bb" />

#### Chat

<img width="914" height="500" alt="Image" src="https://github.com/user-attachments/assets/f07a7a4c-9c61-4548-a280-59a7d4bcdb2d" />

I looked at the logits of the top 10 tokens, and it seems like there is a state transition from generating genuine zeros to getting stuck at generating zeros forever. Darker lines represent later tokens:

<img width="1259" height="1017" alt="Image" src="https://github.com/user-attachments/assets/a0c6a229-02d3-4df7-85c3-4a8cd7457af1" />

I am not sure whether it is sufficient to look at just the logits for reliable detection, but maybe there are some infinite repeat neurons in the network that can be identified and disabled.

However, it would be more desirable to fix this issue during training. Here are a few ideas:

It would probably help to train on sequences with lots of repetition that eventually transition into non-repeating sequences, so the model can learn to break out of those loops. Maybe the issue lies in the way that the training data has been prepared. Commonly, there is a preprocessing step where "low quality" data is removed. Perhaps all data with lots of repetition has been identified as low quality data and been removed, so the model has never learnt how to deal with it.

Another approach would be to generate token coordinates in addition to tokens. When too many tokens are generated at the same coordinates, the model is obviously stuck, which is much easier to detect than repeating text sequences. It is also easy to penalize this kind of generation with an appropriate loss function during training. Likewise, coordinates outside of the image boundaries are easy to penalize as well.

Alternatively, the model could indicate the image patch from which the currently generated token originated. There are only so many characters that can fit into an image patch, so it should be easy to introduce a limit here. It might make sense to combine this with some kind of weight based on font size, since larger characters take up more space in a patch.

**Comment by scouzi1966** [comment]
@99991 It actually passes on my app with Qwen3-VL-8B (fp16).  Backend in this case is llama.cpp with Apple metal

<img width="1643" height="466" alt="Image" src="https://github.com/user-attachments/assets/13da4247-64bf-4504-abe9-c02af96dcdee" />

**Comment by 99991** [comment]
@scouzi1966 It is nice that it passes for you on a larger model with llama.cpp backend, but it would be nice if it worked all the time on all images, not just sometimes.

I have begun work on [a benchmark](https://github.com/99991/LoopyLeaderboard) to quantify the repetition issue.

| Loopy Score | # Loopy Outputs | Decoding | Model |
| - | - | - | - |
|   0.0 % |   0 | greedy     | [GLM-4.6V-Flash-Q5_K_M](https://huggingface.co/unsloth/GLM-4.6V-Flash-GGUF/blob/main/GLM-4.6V-Flash-Q5_K_M.gguf) |
|   2.8 % |  14 | greedy     | [gemma-3-27b-it-Q4_K_M](https://huggingface.co/ggml-org/gemma-3-27b-it-GGUF/blob/main/gemma-3-27b-it-Q4_K_M.gguf) |
|   4.4 % |  22 | greedy     | [Mistral-Small-3.1-24B-Instruct-2503-Q4_K_M](https://huggingface.co/unsloth/Mistral-Small-3.1-24B-Instruct-2503-GGUF/blob/main/Mistral-Small-3.1-24B-Instruct-2503-Q4_K_M.gguf) |
|   8.0 % |  40 | greedy     | [gemma-3-4b-it-Q4_K_M](https://huggingface.co/ggml-org/gemma-3-4b-it-GGUF/blob/main/gemma-3-4b-it-Q4_K_M.gguf) |
|  21.4 % | 107 | greedy     | [Qwen3-VL-30B-A3B-Instruct-Q4_K_S](https://huggingface.co/unsloth/Qwen3-VL-30B-A3B-Instruct-GGUF/blob/main/Qwen3-VL-30B-A3B-Instruct-Q4_K_S.gguf) |
|  24.4 % | 122 | greedy     | [Qwen3-VL-8B-Instruct-Q5_K_M](https://huggingface.co/unsloth/Qwen3-VL-8B-Instruct-GGUF/blob/main/Qwen3-VL-8B-Instruct-Q5_K_M.gguf) |
|  30.2 % | 151 | greedy     | [InternVL3-14B-Instruct-Q4_K_M](https://huggingface.co/ggml-org/InternVL3-14B-Instruct-GGUF/blob/main/InternVL3-14B-Instruct-Q4_K_M.gguf) |
|  36.0 % | 180 | greedy     | [Qwen3-VL-8B-Instruct](https://huggingface.co/Qwen/Qwen3-VL-2B-Instruct) |
|  36.4 % | 182 | greedy     | [Nanonets-OCR-s-Q4_K_M](https://huggingface.co/unsloth/Nanonets-OCR-s-GGUF/resolve/main/Nanonets-OCR-s-Q4_K_M.gguf) |
|  39.6 % | 198 | non-greedy | [Qwen3-VL-4B-Instruct](https://huggingface.co/Qwen/Qwen3-VL-4B-Instruct) |
|  40.8 % | 204 | greedy     | [Qwen3-VL-4B-Instruct-BF16](https://huggingface.co/unsloth/Qwen3-VL-4B-Instruct-GGUF/blob/main/Qwen3-VL-4B-Instruct-BF16.gguf) |
|  40.8 % | 204 | greedy     | [Qwen3-VL-4B-Instruct-Q5_K_M](https://huggingface.co/unsloth/Qwen3-VL-4B-Instruct-GGUF/blob/main/Qwen3-VL-4B-Instruct-Q5_K_M.gguf) |
|  41.8 % | 209 | non-greedy | [Qwen3-VL-2B-Instruct](https://huggingface.co/Qwen/Qwen3-VL-2B-Instruct) |
|  45.8 % | 229 | greedy     | [Qwen3-VL-4B-Instruct](https://huggingface.co/Qwen/Qwen3-VL-4B-Instruct) |
|  49.8 % | 249 | greedy     | [Qwen3-VL-2B-Instruct](https://huggingface.co/Qwen/Qwen3-VL-2B-Instruct) |
|  54.6 % | 273 | greedy     | [Qwen3-VL-2B-Instruct-BF16](https://huggingface.co/unsloth/Qwen3-VL-2B-Instruct-GGUF/blob/main/Qwen3-VL-2B-Instruct-BF16.gguf) |

Insights so far:

* Greedy decoding leads to more repetition
* Larger models are better
* Gemma models are much more robust than Qwen models

**EDIT**

I have just benchmarked GLM-4.6V-Flash, which is absolutely perfect and has zero repetition for all 500 image and prompt combinations I tested.

**Comment by nitin302** [comment]
seeing the same issue

**Comment by scruffynerf** [comment]
+1, annoying, no solution?

**Comment by AKCqhzdy** [comment]
So I wander whether there are any available methods to avoid repetition? In the production environment partial absence of results should be allowed more than continuous repetition, since it may lead to a significant consumption of time and space

**Comment by BruceHuQi** [comment]
I use the Qwen3-VL-4B model fine-tuned on A100 for furniture recognition on multiple images with a 4090 GPU, but some of the images yield duplicate outputs.

**Comment by moarseniy** [comment]
I have the same issue for bounding box detection using Qwen3-VL-32B. 
Repeated same bounding boxes in output.

**Comment by 99991** [comment]
I have just benchmarked the [GLM-4.6V-Flash](https://huggingface.co/zai-org/GLM-4.6V-Flash) model using [llama.cpp](https://github.com/ggml-org/llama.cpp/) with [Q5_K_M quantization](https://huggingface.co/unsloth/GLM-4.6V-Flash-GGUF/tree/main), which is absolutely perfect and had zero repetition for all 500 image and prompt combinations I tested. I have also OCRed an additional 500 PDF pages from arXiv papers and could not find any major issues. It required less than 11 GB VRAM and you can probably fit it in under 8GB with stronger quantization. I'd recommend everyone to check whether GLM-4.6V-Flash can replace Qwen3-VL. I, at least, am extremely happy with the model!
